### ISSUE TITLE
.loc is incorrect  for compiler generated spill code in ppc32/64 .

### ISSUE THREAD

**Issue by UmeshKalappa0** [llvm/llvm-project#56370]
Hi ,
The testcase like https://godbolt.org/z/McGWf7YE8 .

.loc is should be  associated like ".loc 0 0" for the compiler generated insn like 

 .loc    1 25 4                          # example.c:25:4
        bl callee
        lwz 3, 8(31)                            # 4-byte Folded Reload

it should be like 
 .loc    1 25 4                          # example.c:25:4
   bl callee
.loc 1 0 0 
   lwz 3, 8(31)                            # 4-byte Folded Reload

instead of previous stmt .

problem pops up when we nexti from the gdb or any dwarf consumer.

**Comment by pogo59** [comment]
This is curious, because https://reviews.llvm.org/D127488 is attempting to eliminate line-0 locations, on the grounds that they take up space and aren't useful for debugging.

> problem pops up when we nexti from the gdb or any dwarf consumer.

What problem is it that you see?

**Comment by chenzheng1030** [comment]
> What problem is it that you see?

Yeah, same question here. I guess `nexti` should not rely on debug line section, it just executes the instructions one by one. So inserting a `.loc 1 0 0` between the call and the `lwz`(inserted internally by compiler) will not impact the `nexti`. And it will also not impact the `next` as I tested on PPC linux with GDB.

**Comment by UmeshKalappa0** [comment]
>>What problem is it that you see?
the issue we have like 

step to callee ,then return from callee like finish ,we expect gdb to go .loc 25 not to stay in .loc 24 .

**Comment by dwblaikie** [comment]
Is there precedent for this in other compilers (GCC)?
I'd generally err away from this and suggest that this sort of behavior (unlocated instructions "attaching" to the previous located instruction) is to be expected from most producers.
I guess associating them with the next line instead could improve user experience, though might also produce confusion too (eg: get to the call line, nexti and nothing happens/the call doesn't execute because some extra instructions are there (though it wouldn't be super surprising, since many/most calls need some setup instructions to put arguments in the registers, etc)) - and also that'd risk breaking things if you back propagated the location over a basic block boundary, which would mean potentially making a more inconsistent user experience (or a larger line table by using line zero in that special case).

**Comment by pogo59** [comment]
> that'd risk breaking things if you back propagated the location over a basic block boundary,

We already go to some trouble to avoid doing that, due to user complaints.  There's a high chance the previous block is completely unrelated and will only cause more confusion.

Spills and reloads are, in one view, completely artificial and there's an argument that they should therefore get line 0.  On the other hand, spills and reloads are there primarily because the calling convention says certain registers are caller-preserved, and arguably the spill/reload exists only because there's a call, suggesting that they really should be attached to the same source location as the call.

**Comment by UmeshKalappa0** [comment]
@pogo59 and @chenzheng1030  ,agree that spills and reloads are part of call and location of those insn should be associated with call .loc .

Now ,says we step over the call and step over lands you to these insn because your return address update same ,and i see the problem here (from the  perspective of gdb ).
